In [25]:
import pandas as pd

In [26]:
course_data = pd.read_csv('../src/data/courses.csv')
course_data.head()

url change changeDescription  \
0  https://selcrs.nsysu.edu.tw/menu5/showoutline....     新增              7/15   
1  https://selcrs.nsysu.edu.tw/menu5/showoutline....    NaN               NaN   
2  https://selcrs.nsysu.edu.tw/menu5/showoutline....     異動          更換教室9/16   
3  https://selcrs.nsysu.edu.tw/menu5/showoutline....    NaN               NaN   
4  https://selcrs.nsysu.edu.tw/menu5/showoutline....    NaN               NaN   

   multipleCompulsory department      id  grade class  \
0               False       中學學程  STP101      0   不分班   
1               False       中學學程  STP104      0   不分班   
2               False       中學學程  STP117      0   不分班   
3               False       中學學程  STP118      0   不分班   
4               False       中學學程  STP120      0   不分班   

                                         name  credit  ... selected  \
0               教育心理學\nEDUCATIONAL PSYCHOLOGY     2.0  ...       37   
1                教學原理\nPRINCIPLES OF TEACHING     2.0  ...       47   
2                  班級經營\nCLASSROOM MANAGEMENT     2.0  ...       47   
3  課程發展與設計\nCURRICULUM DEVELOPMENT AND DESIGN     2.0  ...       23   
4               青少年心理學\nADOLESCENT PSYCHOLOGY     2.0  ...       37   

   remaining  teacher            room                       classTime  \
0         13      馮雅群  三5,6(社SS 2001)  ['', '', '56', '', '', '', '']   
1          3      張至慶  四6,7(社SS 2006)  ['', '', '', '67', '', '', '']   
2          3      陳利銘  四3,4(社SS 2001)  ['', '', '', '34', '', '', '']   
3         27      周珮儀  二6,7(社SS 2006)  ['', '67', '', '', '', '', '']   
4         13      邱文彬  一3,4(社SS 2005)  ['34', '', '', '', '', '', '']   

              description        tags english  \
0       《講授類》\n本課程為教育學程課程          []   False   
1  《講授類》\n本課程為教育專業課程，雙語授課          []   False   
2       《講授類》\n本課程為教育學程課程          []   False   
3       《講授類》\n本課程為教育學程課程  ['環境教育學程']   False   
4       《講授類》\n本課程為教育學程課程          []   False   

                                            syllabus  \
0  本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...   
1  \\140.117.152.55\crs_net_data\crs_outline\1131...   
2  本課程將選擇希臘悲劇、喜劇、莎士比亞悲劇以及中國傳統戲曲中的經典劇目（若課堂時間足夠亦補充日...   
3  本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...   
4  本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...   

                                          objectives  
0  Preview the current topics in the area of info...  
1             1.終身運動觀念的建立2.正確安全地從事運動3.各項競技運動技術的掌握與理解  
2  藉由研讀、觀賞與討論東西方經典劇目了解東西方戲劇的特質差異，進而探究造成差異的深層文化背景，...  
3  This course aims to provide students with a so...  
4  This is a follow-up course of Financial Manage...  

[5 rows x 24 columns]

In [27]:
# Check for empty string in objectives and syllabus
course_data[course_data['objectives'] != '']['objectives']

0       Preview the current topics in the area of info...
1                  1.終身運動觀念的建立2.正確安全地從事運動3.各項競技運動技術的掌握與理解
2       藉由研讀、觀賞與討論東西方經典劇目了解東西方戲劇的特質差異，進而探究造成差異的深層文化背景，...
3       This course aims to provide students with a so...
4       This is a follow-up course of Financial Manage...
                              ...                        
2568    After completing this course the student shoul...
2569                                                依規定免登
2570    This course will introduce various prevalent f...
2571                                                  NaN
2572    1 培養健康的歌唱與發聲習慣與方法2 認識各類合唱音樂之風格3 學習團體歌唱之技巧：如何以聆...
Name: objectives, Length: 2573, dtype: object

本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Provide information of course syllabus in English.(This is for statistical use only. For those who do not provide information of course syllabus in English, do not check this field.)】\\140.117.152.55\crs_net_data\crs_outline\1131\outline\FM203.TXTFinancial accounting is the branch of accounting that focuses on producing information for external users of financial reports, such as shareholders and lenders. The information generated by the financial reporting system is essential for the efficient operation of global financial markets.This course is designed to develop the problem-solving and research skills necessary to understand issues related to the collection, reporting, and analysis of financial accounting information from the preparer’s perspective for decision-making. It aims to provide an in-depth exploration of financial accounting, including the preparation of financial statements involving more complex economic events and the recognition of assets. The course equips students with the knowledge and skills needed to advance to further studies in financial reporting.

本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Provide information of course syllabus in English.(This is for statistical use only. For those who do not provide information of course syllabus in English, do not check this field.)】Introduction to current topics in information management

In [28]:
course_data[course_data['syllabus'] != '']['syllabus']

0       本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...
1       \\140.117.152.55\crs_net_data\crs_outline\1131...
2       本課程將選擇希臘悲劇、喜劇、莎士比亞悲劇以及中國傳統戲曲中的經典劇目（若課堂時間足夠亦補充日...
3       本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...
4       本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...
                              ...                        
2568    1. Introduction and Microstructure2. Polymer S...
2569                                                依規定免登
2570    本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...
2571    本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）【Pr...
2572              訓練演唱能力、聽力、團體歌唱技巧以及鑑賞力，並擴展對不同時代曲目的詮釋力能力。
Name: syllabus, Length: 2573, dtype: object

In [29]:
# Clean the data
course_data['syllabus'] = course_data['syllabus'].apply(lambda x: str(x).replace('【Provide information of course syllabus in English.(This is for statistical use only. For those who do not provide information of course syllabus in English, do not check this field.)】', ''))

course_data['syllabus'] = course_data['syllabus'].apply(lambda x: str(x).replace('本課程教學大綱已提供完整英文資訊（本選項僅供統計使用，未提供完整英文資訊者，得免勾記）', ''))

course_data[course_data['syllabus'] != '']['syllabus']

0       Introduction to current topics in information ...
1       \\140.117.152.55\crs_net_data\crs_outline\1131...
2       本課程將選擇希臘悲劇、喜劇、莎士比亞悲劇以及中國傳統戲曲中的經典劇目（若課堂時間足夠亦補充日...
3       \\140.117.152.55\crs_net_data\crs_outline\1131...
4       \\140.117.152.55\crs_net_data\crs_outline\1131...
                              ...                        
2568    1. Introduction and Microstructure2. Polymer S...
2569                                                依規定免登
2570    (1) ETF market(2) Foreign exchange(3) Digital ...
2571                Independent study of master students.
2572              訓練演唱能力、聽力、團體歌唱技巧以及鑑賞力，並擴展對不同時代曲目的詮釋力能力。
Name: syllabus, Length: 2573, dtype: object

In [30]:
import re

def clean_syllabus(text: str) -> str:
    # Step 1: Remove file paths
    text = re.sub(r'\\\\[^\s]+\.TXT', '', text)

    # Step 2: Remove leading numeric indices (e.g., "1,0," or "2,1,")
    text = re.sub(r'^\d+,\d+,?', '', text)

    # Step 3: Strip any surrounding whitespaces
    text = text.strip()
    
    return text

course_data['syllabus'] = course_data['syllabus'].apply(clean_syllabus)

course_data[course_data['syllabus'] != '']['syllabus']

0       Introduction to current topics in information ...
1       1.基本認知：各項競技運動規則、熱身方法、傷害預防與簡易處理2.實際演練：基本技術的練習、結...
2       本課程將選擇希臘悲劇、喜劇、莎士比亞悲劇以及中國傳統戲曲中的經典劇目（若課堂時間足夠亦補充日...
3       Financial accounting is the branch of accounti...
4       Investment decisionInvestmentsFinancingCapital...
                              ...                        
2568    1. Introduction and Microstructure2. Polymer S...
2569                                                依規定免登
2570    (1) ETF market(2) Foreign exchange(3) Digital ...
2571                Independent study of master students.
2572              訓練演唱能力、聽力、團體歌唱技巧以及鑑賞力，並擴展對不同時代曲目的詮釋力能力。
Name: syllabus, Length: 2573, dtype: object